src/shims/fs.rs

```diff
@@ -160,20 +171,31 @@ pub trait EvalContextExt<'mir, 'tcx: 'mir>: crate::MiriEvalContextExt<'mir, 'tcx
             return Ok(0);
         }
         let fd = this.read_scalar(fd_op)?.to_i32()?;
-        let buf_scalar = this.read_scalar(buf_op)?.not_undef()?;
-
-        // Remove the file handle to avoid borrowing issues.
-        this.remove_handle_and(fd, |mut handle, this| {
-            // Don't use `?` to avoid returning before reinserting the handle.
-            let bytes = this.force_ptr(buf_scalar).and_then(|buf| {
-                this.memory
-                    .get_mut(buf.alloc_id)?
-                    .get_bytes_mut(&*this.tcx, buf, Size::from_bytes(count))
-                    .map(|buffer| handle.file.read(buffer))
-            });
-            this.machine.file_handler.handles.insert(fd, handle).unwrap_none();
-            this.try_unwrap_io_result(bytes?.map(|bytes| bytes as i64))
-        })
+        let buf = this.read_scalar(buf_op)?.not_undef()?;
+
+        if let Some(handle) = this.machine.file_handler.handles.get_mut(&fd) {
+            let count = helpers::try_into_host_usize(count)
+                .ok_or_else(|| err_unsup_format!("Program tries to read into buffer too big for this host platform"))?;
+            // We want to read at most `count` bytes
+            let mut bytes = vec![0; count];
+            let result = handle.file.read(&mut bytes);
+
+            match result {
+                Ok(c) => {
+                    let read_bytes = helpers::try_from_host_usize::<i64>(c)
+                        .ok_or_else(|| err_unsup_format!("Number of read bytes {} cannot be transformed to i64", c))?;
```

Comment:
So... here we convert to `i64` and then, *later*, we convert to what the target platform really needs. What will happen if that second part is the one that overflows? It's probably a `from_int` in `foreign_items.rs`, and that will ICE if the result is too big. So, the error handling isn't really adequate.

But also, do we need this? We know `c <= count`. What is the largest possible `count`? You used `to_usize`, is that right? It seems the type in C is `size_t`, which would be `isize`, not `usize`! So we should know that the buffer is definitely not too big to represent the result.

The only place where we can actually error is if the program does sth too big for the target, but that is already handled above.

Reply:
Ah no, `size_t` *is* unsigned, I was wrong there. But that's silly, what does the original function do if the size fits into a `size_t` but not an `ssize_t`?!? I hate C...

Reply:
@oli-obk any opinions? My conclusion is that the API we are trying to emulate is inherently broken and there is nothing reasonable we can do, so we should do something simple and consistent.^^

We could interpret the `count` parameter as signed, so if it is "too big" it'll be negative, and then we usize `usize::try_from` and if that complains we throw an "unsupported big buffer" or so?

Reply:
Since the docs on `read` say

> Upon successful completion, read() and pread() shall return a non-negative integer indicating the number of bytes actually read. Otherwise, the functions shall return -1 and set errno to indicate the error. 

I would say that any value larger than `isize::max_value()` (for the target `isize`) should throw an interpreter error. If the target is 64 bit, it can't happen anyway, because we can't have vectors larger than `isize::max_value()`.

Reply:
Well, these functions don't have to fill the entire buffer. So I assume what happens is, they truncate the input to `isize::MAX` and will never use more than that.

Reply:
right, so the `usize` we get from Rust's `Read::read` will never be > i64::max_value(), so for 64bit host platforms interpreting 32 bit target platforms, we basically have to give up when such a situation occurs, since it's too late to do anything about it. Everywhere else, it should be fine

Reply:
If we truncate the buffer size we pass to `Read::read` we know the return value can't be bigger than that.

Reply:
Let me see if IIUC. So that basically means being sure that the `count` parameter doesn't exceeds `isize::max_value()` so that when we create the intermediate buffer for reading/writing, it isn't larger than `isize::max_value()` and we can return a proper value at the end. Sounds good to you?

Reply:
I think I figured it out for `read`. I'm not sure if `write` is OK tho.